### PR TITLE
Test with rails 7

### PR DESCRIPTION
**Pull request by Intrepidd**
Wanted to see if rails 7 could be easily supported, but I'm getting a lot of failures

--------
Thank you for your contribution!

Check the following boxes:

- [ ] Wrote [good commit messages][1].
- [ ] Feature branch is up-to-date with `master` (if not - rebase it).
- [ ] Squashed related commits together.
- [ ] Added tests.
- [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new 
  code introduces user-observable changes.
- [ ] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by dhekimian** [comment]
The only error I see is that Rails 7 depends on Ruby (>=2.7.0) in the `gha-workflow-pt-test / Ruby 2.6, rails_7.0.gemfile (pull_request)` test. 

>  rails (~> 7.0.0) was resolved to 7.0.0, which depends on
      Ruby (>= 2.7.0)

Are there other errors your seeing?

**Comment by Intrepidd** [comment]
Yes it's not showing because most builds got cancelled after a timeout but I'm getting many failures 👍 

```
Finished in 9.8 seconds (files took 2.68 seconds to load)
411 examples, 21 failures

Failed examples:

rspec ./spec/models/document_spec.rb:15 # Document#paper_trail.next_version returns the expected document
rspec ./spec/paper_trail/compatibility_spec.rb:16 # PaperTrail::Compatibility.check_activerecord when incompatible writes a warning to stderr
rspec ./spec/paper_trail/version_limit_spec.rb:50 # PaperTrail::Cleaner deletes oldest versions, when the database returns them in a different order
rspec ./spec/models/widget_spec.rb:587 # Widget#next_version with a reified item returns the object (not a Version) as it became next
rspec ./spec/models/widget_spec.rb:259 # Widget with a persisted record when updated, and then destroyed have nil item for last version
rspec ./spec/models/widget_spec.rb:101 # Widget with a persisted record when updated with changes be available in its previous version
rspec ./spec/models/widget_spec.rb:503 # Widget with an item with versions .versions_between return versions in the time period
rspec ./spec/models/widget_spec.rb:489 # Widget with an item with versions when the versions were created over time still return a widget when appropriate, when passing timestamp as string
rspec ./spec/models/widget_spec.rb:614 # Widget#previous_version with a reified item returns the object (not a Version) as it was most recently
rspec ./spec/models/widget_spec.rb:320 # Widget with a record's papertrail handle datetimes
rspec ./spec/models/widget_spec.rb:333 # Widget with a record's papertrail handle booleans
rspec ./spec/models/widget_spec.rb:308 # Widget with a record's papertrail handle integers
rspec ./spec/models/widget_spec.rb:300 # Widget with a record's papertrail handle strings
rspec ./spec/models/widget_spec.rb:304 # Widget with a record's papertrail handle text
rspec ./spec/models/widget_spec.rb:316 # Widget with a record's papertrail handle decimals
rspec ./spec/models/widget_spec.rb:324 # Widget with a record's papertrail handle times (time only, no date)
rspec ./spec/models/widget_spec.rb:312 # Widget with a record's papertrail handle floats
rspec ./spec/models/widget_spec.rb:329 # Widget with a record's papertrail handle dates
rspec ./spec/models/widget_spec.rb:344 # Widget with a record's papertrail when a column has been removed from the record's schema restore all forward-compatible attributes
rspec ./spec/models/widget_spec.rb:340 # Widget with a record's papertrail when a column has been removed from the record's schema reify previous version
rspec ./spec/models/animal_spec.rb:17 # Animal works with custom STI inheritance column
```

You can check it out by running `DB=sqlite bundle exec appraisal rails-7.0 rake`

**Comment by Intrepidd** [comment]
I ignored the rails 7 + ruby 2.6 run

Not sure why rails <7 builds are getting canceled but specs seem to pass before the canceling happens

On rails 7 though there are several errors : https://github.com/paper-trail-gem/paper_trail/runs/4581204133?check_suite_focus=true

I tried looking them up but I'm not familiar enough with the code, maybe someone from the team can look it up ? @jaredbeck  or @batter  ?

**Comment by AlfonsoUceda** [comment]
@Intrepidd it is because of the `fail-fast` config in the github actions. If you set to false inside `strategy` section, you will see the others steps to finish.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -70,6 +70,9 @@ jobs:
           # https://github.com/rails/rails/issues/40938
           - ruby: '3.0'
             gemfile: 'rails_5.2'
+          # rails 7 requires ruby >= 2.7.0
+          - ruby: '2.6'
```
Comment (pedrocarmona) [inline .github/workflows/test.yml:74]:
```suggestion
          - ruby: '2.7'
```

Comment (Intrepidd) [reply .github/workflows/test.yml:74]: This is an exclusion list, so I make sure we don't run tests for ruby 2.6 + rails 7.0

**Comment by pedrocarmona** [comment]
One problem seams to be related with this rails 7 setting `automatic_scope_inversing`.

```ruby
# spec/dummy_app/config/application.rb
   config.active_record.automatic_scope_inversing = false
```

with this flag it fixes 9 errors.

**Comment by AlfonsoUceda** [comment]
@Intrepidd about that @pedrocarmona said, maybe we can "fix" the inverse_of associations with the right name and class. I think that config is 7.0 specific. What do you think?

**Comment by Intrepidd** [comment]
@AlfonsoUceda @pedrocarmona Thanks! Indeed passing `inverse_of: nil` removes 9 errors and they seem to share the same root, I'll investigate those now

**Comment by Intrepidd** [comment]
The issue seems to be related to time deserialization, with rails.7 the deserializer gets an `ActiveRecord::Type::Time::Value` instead of an `ActiveSupport::TimeWithZone`

**Comment by Intrepidd** [comment]
Looks like it comes from this rails commit : https://github.com/rails/rails/commit/f45b01ad3d98bca680fb9912292fea32f453f9da#diff-49895b7c6452971e90d4809004817a390388351d60420be48c4f35c82705b7ef

**Comment by Intrepidd** [comment]
Rails 6 : 

```
(byebug) YAML.dump object["a_time"][1]
"--- !ruby/object:ActiveSupport::TimeWithZone\nutc: 2000-01-01 14:30:27.906201000 Z\nzone: !ruby/object:ActiveSupport::TimeZone\n  name: Etc/UTC\ntime: 2000-01-01 14:30:27.906201000 Z\n"
```

Rails 7 :

```
(byebug) YAML.dump object["a_time"][1]
"--- !ruby/object:ActiveRecord::Type::Time::Value\nutc: 2000-01-01 14:28:40.322147000 Z\nzone: !ruby/object:ActiveSupport::TimeZone\n  name: Etc/UTC\ntime: 2000-01-01 14:28:40.322147000 Z\n"
```

In both cases, the object is an `ActiveRecord::Type::Time::Value`

**Comment by pedrocarmona** [comment]
Adding this statement to the end of the file, patches the problem
```
# lib/paper_trail/serializers/yaml.rb
if ::ActiveRecord.version >= ::Gem::Version.new("7.0.0.alpha")
  ::YAML.dump_tags[ActiveRecord::Type::Time::Value] = "!ruby/object:ActiveSupport::TimeWithZone"
end
```

However, maybe this should be fixed in rails itself.

**Comment by Intrepidd** [comment]
> Adding this statement to the end of the file, patches the problem
> 
> ```
> # lib/paper_trail/serializers/yaml.rb
> if ::ActiveRecord.version >= ::Gem::Version.new("7.0.0.alpha")
>   ::YAML.dump_tags[ActiveRecord::Type::Time::Value] = "!ruby/object:ActiveSupport::TimeWithZone"
> end
> ```
> 
> However, maybe this should be fixed in rails itself.

Yeah this works but this creates a nasty side effect.

Maybe the error is in Rails as the value can't seem to be deserialized from yaml : 

```ruby
# Rails 6.1
(byebug) YAML.load(YAML.dump(ActiveRecord::Type::Time::Value.new(Time.current))) + 1.day
Thu, 23 Dec 2021 10:24:42.873895000 UTC +00:00

# Rails 7
(byebug) YAML.load(YAML.dump(ActiveRecord::Type::Time::Value.new(Time.current))) + 1.day
*** ArgumentError Exception: not delegated
```

I'll open an issue

File: lib/paper_trail/compatibility.rb [hunk 2]
```diff
@@ -18,7 +18,7 @@ module PaperTrail
   # versions.
   module Compatibility
     ACTIVERECORD_GTE = ">= 5.2" # enforced in gemspec
-    ACTIVERECORD_LT = "< 7.0" # not enforced in gemspec
+    ACTIVERECORD_LT = "< 8.0" # not enforced in gemspec
```
Comment (jaredbeck) [inline lib/paper_trail/compatibility.rb:21]: Please specify `< 7.1`

**Review by jaredbeck — COMMENTED** [review]
Hi! Just wanted to drop in and thank you for your work here. I've been watching the conversation casually, and look forward to digging deeper when this PR is ready for review.

File: lib/paper_trail/compatibility.rb (hunk 2 above)
Comment (Intrepidd) [reply lib/paper_trail/compatibility.rb:21]: Could you explain why ? When the latest rails was 6.1 this was `<7.0` and not `< 6.2`, I believe as long as we are in the same major version there should be no breaking change ?

Comment (jaredbeck) [reply lib/paper_trail/compatibility.rb:21]: Yeah, I can see why the previous value would be confusing. The previous value should have been `< 6.2`. The intention of this `Compatibility` range is for versions that we test against. IOW, if it's in the range, we are actively testing against it in CI.

Comment (Intrepidd) [reply lib/paper_trail/compatibility.rb:21]: Got it !

**Comment by AlfonsoUceda** [comment]
Hey folks, thanks for your work.

Is there an ETA to support this in the main branch?

**Comment by AlfonsoUceda** [comment]
@Intrepidd should this PR be marked as ready to review?

**Comment by Intrepidd** [comment]
> @Intrepidd should this PR be marked as ready to review?

Not yet we are still waiting on https://github.com/rails/rails/issues/43966

**Comment by pbstriker38** [comment]
Does this only effect yaml, or is json also effected?

**Comment by Intrepidd** [comment]
> Does this only effect yaml, or is json also effected?

Only YAML AFAICS

**Comment by mathieujobin** [comment]
I think we can close this PR in favor of https://github.com/paper-trail-gem/paper_trail/pull/1365

**Comment by jaredbeck** [comment]
Closing via https://github.com/paper-trail-gem/paper_trail/pull/1365

**Pull request closed without merging**